### PR TITLE
readers/mutlishard: don't mix coroutines and continuations in the do_fill_buffer()

### PR DESCRIPTION
**Pull request by avikivity**
The combination is hard to read and modify.

### CODE REVIEW

File: readers/multishard.cc [hunk 1]
```diff
@@ -803,8 +803,9 @@ future<> shard_reader_v2::do_fill_buffer() {
         remote_fill_buffer_result_v2 result;
     };
 
-    if (!_reader) {
-        fill_buf_fut = smp::submit_to(_shard, [this, gs = global_schema_ptr(_schema)] () -> future<reader_and_buffer_fill_result> {
```
Comment (denesb) [inline readers/multishard.cc:807]: `fill_buf_fut` now seems to be unused, should be removed.

File: readers/multishard.cc [hunk 2]
```diff
@@ -833,20 +834,22 @@ future<> shard_reader_v2::do_fill_buffer() {
             }
             co_await rreader->close();
             std::rethrow_exception(std::move(ex));
-        }).then([this] (reader_and_buffer_fill_result res) {
+        }));
+        {
```
Comment (denesb) [inline readers/multishard.cc]: Why these sub-scopes for returns (there is one more below)?

File: readers/multishard.cc [hunk 3]
```diff
@@ -833,20 +834,22 @@ future<> shard_reader_v2::do_fill_buffer() {
...
+      } else {
+        co_return co_await smp::submit_to(_shard, coroutine::lambda([this] () -> future<remote_fill_buffer_result_v2>  {
             reader_permit::used_guard ug{_reader->permit()};
-            return _reader->fill_buffer().then([this, ug = std::move(ug)] {
-                return remote_fill_buffer_result_v2(_reader->detach_buffer(), _reader->is_end_of_stream());
-            });
-        });
-    }
+            co_await _reader->fill_buffer();
+            {
```
Comment (denesb) [inline readers/multishard.cc]: here too

**Review by denesb — COMMENTED** [review]
Looks good, small comments only.

File: readers/multishard.cc (hunk 2 above)
Comment (avikivity) [reply readers/multishard.cc]: It's to preserve indentation. Removed in the next patch..

File: readers/multishard.cc (hunk 1 above)
Comment (avikivity) [reply readers/multishard.cc:807]: yes

**Pull request closed without merging**